DOC-3246: Suggested Edits and User Lookup API documentation (#3798)
tinymce/tinymce-docs: modules/ROOT/pages/suggestededits.adoc

= Suggested Edits Plugin
:navtitle: Suggested Edits 
:description: Suggested Edits keeps track of changes and review edits made by multiple authors in a document.
:description_short: A view of tracked changes made by multiple authors

> :keywords: plugin, Suggested Edits, changes, diff, track changes, review changes, collaboration, multiple authors

Author: I think it'd be good to have it appear if the search is "track changes"

:plugincode: suggestededits
:plugincategory: premium

include::partial$misc/admon-paid-addon-pricing.adoc[]

The {pluginname} plugin allows multiple users to collaborate on a document. The plugin shows which user suggested which edits, whether they added, removed, modified, or replaced any content, and allows users to provide feedback on those suggestions or give a final review by approving or rejecting them.

== Interactive example

liveDemo::{plugincode}[]

include::partial$misc/admon-iframe-only.adoc[]

== How it works

The {pluginname} plugin keeps track of every suggested edit made to the document by the current user, as the edits are made, and stores this metadata in an internal model of the document. These suggestions can then be reviewed in the `'suggestededits'` xref:custom-view.adoc[editor view], where each edit is highlighted in the document, and where you can accept, reject, or provide feedback. The `'suggestededits'` view is accessible via either the `suggestededits` toolbar button or menu button within the `View` menu. These edits, and any feedback provided, are stored in the model provided by the plugin xref:#get_model[`+get_model()+` API].

[IMPORTANT]

> This model should be saved externally alongside the document, and loaded into the editor with the xref:#suggestededits_model[`+suggestededits_model+` option]. This will ensure that both the document and the model are in sync, and that every user's contributions are tracked correctly.

Reviewer: Visually any text rendered inside `backticks` that is rendered as a link such as `[`+suggestededits_model+` option]`  looks unnatural to me personally, but I'm happy to leave as we don't have a specific standard across the documentation from what i can see atm.

=== Header

The view contains a few controls to manage the review process: 

* Show edits: A toggle button which hides/unhides the highlighted annotations. 
* Complete review: Ends the review, applying the resolved changes to the document.
* Cancel: Exits the review with no change to the document.

=== Document

The current editor document is displayed in a sandboxed iframe, with each suggested edit highlighted as they appear in the document. The following color coding is used to indicate the type of change:  

* Green: Added content. 
* Blue: Modified attributes or formatting (e.g. bold, italic, etc.).
* Red: Removed content.

Replaced content is represented as both added and removed content, and indicates that some content was removed and replaced with new content in a single edit. When a suggestion is selected, whether in the document or in the sidebar, the corresponding highlighted suggestion in the document will be outlined in blue and scrolled into view.

=== Sidebar

> Each suggested edit is listed as a card in the sidebar and color coded by the type of change, along with the user who made the suggestion, when the edit was made, and any feedback provided on that suggestion. When selected, each suggestion can be handled in the following ways:

Author: Not really necessary.

> * Accept: Resolves the edit and will be applied to the document when .
> * Reject: Removes the change from the document.
> * Revert: Undoes a previously accepted or rejected change.
> * Provide feedback: Opens a text area for you to provide feedback on the suggestion.

Author: ```suggestion
* Accept: Resolves the suggestion, applying the edit to the document when the review is completed.
* Reject: Resolves the suggestion, turning back the edit to the original state.
* Revert: Reverts the current "Accept" or "Reject" resolution on the suggestion.
* Provide feedback: Opens a text area for you to provide feedback on the suggestion.
```
I think I just forgot to finish this part up.

The model should be retrieved from the plugin using the xref:#get_model[`+getModel+`] API, then saved externally alongside the document, and loaded into the editor with the xref:#suggestededits_model[`+suggestededits_model+`] option each time the document is opened. This ensures that both the document and the model are in sync, and that every user's contributions are tracked correctly. If the model and content are out of sync, the difference between them will be applied as a suggested edit by the current user. If a model is not provided in the editor configuration or is set to `+undefined+`, the plugin will generate a new model from the initial content.

When the xref:#suggestededits_content[`+suggestededits_content+`] option is set to `'model'`, only the model is required, so keeping a synchronised document is not necessary. This also means that there will be no accidental suggestions caused by the editor content being out of sync with the model.

== Reviewing edits

> The {pluginname} plugin provides a dedicated `'suggestededits'` xref:custom-view.adoc[editor view] for viewing and reviewing edits made by multiple authors. The available actions in this view depends on the xref:#suggestededits_access[`+suggestededits_access+` option].

Reviewer 2: ```suggestion
The Review Edits view can be used to view and reviewing edits made by multiple authors. The available actions in this view depends on the xref:#suggestededits_access[`+suggestededits_access+` option].
```

=== Sidebar

Each suggested edit is listed as a card in the sidebar and color coded by the type of change, along with the user who made the suggestion, when the edit was made, and any feedback provided on that suggestion. When selected, each suggestion can be handled in the following ways:

* Accept: Resolves the suggestion, applying the edit to the document when the review is completed.
* Reject: Resolves the suggestion, turning back the edit to the original state.
* Revert: Reverts the current "Accept" or "Reject" resolution on the suggestion.
* Provide feedback: Opens a text area for you to provide feedback on the suggestion.

Feedback is shown in chronological order beneath the card details, when the card is selected. Feedback threads allows users to discuss suggestions before resolving them. The feedback author can edit or delete their own feedback, with the appropriate permissions in the xref:#suggestededits_access[`+suggestededits_access+`] option.

> At the top of the sidebar, there is a dropdown menu to apply review actions in bulk to all suggested edits:

Reviewer 2: ```suggestion
At the top of the sidebar there is a dropdown menu to apply review actions in bulk to all suggested edits:
```

* Accept all.
* Reject all.
* Revert all.

=== Finishing a review

> When completing a review, any resolved suggestions will be applied to the document and will no longer be tracked in the model as a suggestion, including any feedback on "resolved" suggestions. Any "accepted" edits will remain in the document, and any "rejected" edits will be reverted to the state before the suggestion was made.

Reviewer 2: ```suggestion
When completing a review, resolved suggestions will be applied to the document and will no longer be tracked in the model as a suggestion. Feedback on resolved suggestions is discarded. Any "accepted" edits will remain in the document, and any "rejected" edits will be reverted to the state before the suggestion was made.
```
